match2/commons/opponent_display.lua

```diff
@@ -225,7 +226,7 @@ function OpponentDisplay.InlineTeam(props)
 		template = 'default',
 	}))
 		:gsub('DefaultPage', props.team.pageName)
-		:gsub('DefaultName', props.team.displayName)
+		:gsub('DefaultName', Logic.emptyOr(props.team.displayName, '&nbsp;'))
```

Comment:
We should really provide this as a constant somewhere

Reply:
I'm pretty sure it's better to refer to string literals like these directly.

Reply:
Which software engineering principles do you have in mind?

Reply:
1. If we ever wanted to use a different string than `'&nbsp;'` we would have to change it everywhere, instead of in one place
2. We can use the constant name to explain what it is for those unfamiliar with it
3. When it's a string literal everywhere typos will go undetected, whereas a typo in a constant will cause an exception (which is good)

Basically the same arguments as all magic strings

Reply:
Those are good points (esp #2). But I think it's at too high of a cost, in that using variables for magic strings/numbers makes the code more difficult to read.

Consider a more general example (I'll get to `&nbsp;` later on)

```
local scoreNode = mw.html.create('div')
    :addClass('bracket-score')
    :css('min-width', '300px')
```

The equivalent that uses variables in lieu of magic strings/numbers, and allows for easier rename and spellcheck (for #1 and #3) is: 

```
local MediaWikiMethods = require('Module:MediaWikiMethods') -- Makes it easier to rename methods in the future
local MediaWikiNamespace = require('Module:MediaWikiNamespace') -- Makes it easier to rename mw.html in the future
local CSS_MIN_WIDTH = 'min-width'
local BRACKET_SCORE_MIN_WIDTH = 300
local PX = 'px'
local DIV = 'div'
local BRACKET_SCORE = 'bracket-score'

...

local scoreNode = mw[MediaWikiNamespace.HTML][MediaWikiMethods.CREATE](DIV)
    :addClass(BRACKET_SCORE)
    :css(CSS_MIN_WIDTH, BRACKET_SCORE_MIN_WIDTH .. PX)
```

I think the downsides are clear here. 
- It's harder to read because it's overly verbose (12 lines vs 3), and because it's spread across 4 locations in 3 modules. 
- It's harder for a reader to know that the min width is 300px, since that requires either scrolling to the top of the file, or mousing over the variable if using intellisense. 
- There are now two different ways to refer to the same thing (`DIV` and `'div'`), and a reader might be confused about what the difference between the two are. 
- If `'div'` is used in more than one module, then it's likely that the variable names will differ (for example `DIV` in one module and `HTML_DIV` in another), unless one takes steps to make them conform (say by setting up another module `require('Module:HtmlConstants').DIV`).
- Typography wise, text in all caps is more difficult to read compared to lower case. (This is why outside of code, all caps is used sparingly, for emphasis or when shouting.)




To address your points specifically:
> 1. If we ever wanted to use a different string than '&nbsp;' we would have to change it everywhere, instead of in one place

I don't think this works. For example consider 
```
local NBSP = '&nbsp;'
```
If one wanted to change `&nbsp;` to `&emsp;`, then one would also have to rename the variable name `NBSP` to `EMSP`, unless one is ok with having the variable name stay out of sync with its value. This requires a search and replace. Further more this only works within a module, unless one also sets up a global module `Module:HtmlEntity` and access it via `require('Module:HtmlEntity').NBSP/EMSP`.

I believe that in general, regex search and replace is sufficient for most code transformations. No need to set up dedicated variables for it.

> 2. We can use the constant name to explain what it is for those unfamiliar with it

I agree, having to explain a string/number literal is a good reason to promote it to a separate variable. 

In this case I don't think `&nbsp;` needs explanation. It's expected that people working on this code understand what html entities are, and what non-breaking space does. In any case one can easily look it up. https://www.google.com/search?q=nbsp

> 3. When it's a string literal everywhere, typos will go undetected, whereas a typo in a constant will cause an exception (which is good)

I assume you meant luacheck? Typos in a variable name will only cause exceptions if its nil value is used where nils are not expected. 

With both luacheck and nil exceptions, typos are only caught in the variable name, not its value. For example
```
local DIV = 'diiv' -- Typo
```
will not be caught.

Also luacheck (as currently set up) does not work across module boundaries. So 
```
local HtmlTags = require('Module:HtmlTags')
mw.html.create(HtmlTags.DIIV) -- Typo
```
will not be caught.

Intellisense should eliminate most typos, assuming that I understand what it does. Because any string literal more than a few characters long will be filled out via tab completion, there's not much possibility of a typo.

Reply:
I don't think strings should never again be used inline, but in this specific case, I insist